### DISCUSSION TITLE
Eth mempool

### DISCUSSION THREAD

**Discussion by brianmcgee** [41north/tethys#10] [Ideas]
We can create a consolidated view of the mempool in a KV store within NATS.

Some things to consider:

* each sidecar subscribes to pending transactions and publishes them into a JetStream subject with an appropriate `Msg-Id` header. The subject would be configured to de-duplicate anything received within the last few mins based on that id. 
* we need to encapsulate the tx in an envelope object indicating insertion or removal
* a separate process would then be responsible for consuming the JetStream subject and updating a KV store

### Questions

* Would it be possible to somehow use NATS to co-ordinate with the proxies and elect one of them exclusively as responsible for maintaining the mempool KV store, with re-assignment if a proxy shuts down or is killed? Could we abuse the Queue Group mechanism somehow? This would remove the need for an independent process maintaining the mempool and simplify deployment.

**Comment by brianmcgee** [comment]
In effect we create an aggregated mempool. If you had multiple clients spread across data centers you end up with a very natural global mem pool.

**Comment by brianmcgee** [comment]
If we are unable to use NATS to somehow co-ordinate the proxies and elect one as being responsible for managing the mempool kv store, we can just create a mempool command instead. It would be responsible for listening to all chains and maintaining multiple kv stores. Still quite efficient. 

Plus you can elect to listen to the aggregate feed without running the mempool process if you want and process the updates yourself.

With the use of `MaxConsumers` on the stream we can guarantee only one mempool process can be consuming at a time.

**Reply by aldoborrero** [reply]
Just in case, remember the mempool consists in:

- Broadcasted pending `TXs` coming from the Network (i.e., other peer neighbour shares them).
- Pending `TXs` received to that particular node (which will be exposed to the Network eventually).

For the particular scenario of having one publisher broadcast mempool `TXs`, you're discarding those pending `TXs` that may arrive at a specific node. Of course, eventually, any pending `TX` sent to a particular node will appear in the mempool in the same way as block propagation (but I think there's an obvious advantage to making a distinction here).

**Reply by brianmcgee** [reply]
I'm not suggesting we have one publisher listening to a single client and publishing the transactions. I'm suggesting all sidecars listen to the pending txs subscription for their client and publish into a JetStream defined stream that will handle the de-duplication. 

What I'm talking about above is then processing that stream and updating a KV store.

**Comment by brianmcgee** [comment]
It looks like we can use determinist subject partitioning to ensure any updates related to a specific tx are published within a known subject. This should allow us to distribute the work of maintaining the kv store across all active proxies by ensuring only one of them at a time is responsible for a given key and thereby preserving ordering or processing. 

https://docs.nats.io/nats-concepts/subject_mapping#deterministic-subject-token-partitioning

**Reply by aldoborrero** [reply]
OK, I haven't read so far when I made my first comment above.

This solution is way better as maintaining the global mempool state is composed of the number of nodes you want to use (i.e., increasing the change of seeing pending txs faster).

**Comment by brianmcgee** [comment]
I wonder if a simple queue group would be enough. The real issue would be out of ordering processing of insertions / deletes, however with the de-duplication provided by the stream before processing and the way the mempool works, a first version could just distribute evenly the updates across the proxies and they update the kv store accordingly.
